network/src/p2p/builder.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+use super::*;
+use libp2p::{
+    core::{muxing::StreamMuxerBox, transport, upgrade},
+    futures::{AsyncRead, AsyncWrite},
+    identity,
+    noise::{self, AuthenticKeypair, X25519Spec},
+    pnet::PnetConfig,
+    PeerId, Transport,
+};
+use libp2p_mplex::MplexConfig;
+use std::{borrow::Borrow, time::Duration};
+
+/// Builds libp2p transport for mina with various configurations
+#[derive(Clone)]
+pub struct MinaTransportBuilder {
+    pnet_config: PnetConfig,
+    mplex_config: MplexConfig,
+    noise_keys: AuthenticKeypair<X25519Spec>,
+    timeout: Duration,
+    peer_id: PeerId,
+}
+
+impl MinaTransportBuilder {
+    /// Creates a new instance of [MinaTransportBuilder] with random keypair and empty config
+    pub fn new() -> Self {
+        let keypair = identity::Keypair::generate_ed25519();
+        Self::new_with_key(keypair)
+    }
+
+    /// Creates a new instance of [MinaTransportBuilder] with given keypair and empty config
+    pub fn new_with_key(keypair: identity::Keypair) -> Self {
+        Self::new_with_key_and_config(keypair, MinaTransportConfig::default().borrow())
+    }
+
+    /// Creates a new instance of [MinaTransportBuilder] with given keypair and config
+    pub fn new_with_key_and_config(
+        keypair: identity::Keypair,
+        config: &MinaTransportConfig,
+    ) -> Self {
+        let peer_id = PeerId::from(keypair.public());
+        let shared_key = config.get_shared_key();
+        let pnet_config = PnetConfig::new(shared_key);
+        let mplex_config = config.get_mplex_config();
+        let noise_keys = noise::Keypair::<noise::X25519Spec>::new()
+            .into_authentic(&keypair)
+            .expect("Signing libp2p-noise static DH keypair failed.");
+        Self {
+            pnet_config,
+            mplex_config,
+            noise_keys,
+            timeout: Duration::from_secs(60),
```

Comment:
this also looks like can be moved to TransportConfig and made configurable.

Reply:
There will be static immutable instances for different preset TransportConfig(s), but timeout should be mutable (e.g. via CLI args),